### PR TITLE
on Linux, respect NIX_SSL_CERT_FILE

### LINKED ISSUE #16: Support NIX_SSL_CERT_FILE for a build time trust anchor
NixOS generally install as global ca bundle to `/etc/ssl/certs/ca-bundle.crt`, so in general `ca-certs` works as expected. However test suites are executed in a sandboxed environment where this path does not exist which currently means that we can't run the `cohttp-lwt-unix` test suite.

However, the nix builder environment sets two environment variables if we depend on the ca bundle explicitly. For example:

```
SSL_CERT_FILE=/nix/store/68bpd4p659sm50lb615djysx2znnw4xm-nss-cacert-3.60/etc/ssl/certs/ca-bundle.crt
NIX_SSL_CERT_FILE=/nix/store/68bpd4p659sm50lb615djysx2znnw4xm-nss-cacert-3.60/etc/ssl/certs/ca-bundle.crt
```

It'd be nice if we could support an environment variable like `SSL_CERT_FILE`. This would also be helpful in other situations to override the default bundle location.

### PR DESCRIPTION
**Pull request by hannesm**
fixes #16

### CODE REVIEW

File: lib/ca_certs.ml [hunk 1]
```diff
@@ -85,7 +85,15 @@ let trust_anchors () =
     Bos.OS.Cmd.(run_out cmd |> out_string |> success) >>= function
     | "FreeBSD" -> detect_one freebsd_location
     | "OpenBSD" -> detect_one openbsd_location
-    | "Linux" -> detect_list linux_locations
+    | "Linux" ->
+      begin
+        (* NixOS is special and sets "NIX_SSL_CERT_FILE" as location *)
```
Comment (sternenseemann) [inline lib/ca_certs.ml:90]: This should probably clarify that this is only the case for the sandbox used for building nix packages: At run time on normal NixOS `/etc/ssl/certs/{ca-bundle.crt, ca-certificates.crt}` do both exist and `NIX_SSL_CERT_FILE` is *not* set.

**Review by sternenseemann — CHANGES_REQUESTED** [review]
I think two things can be improved about this:

* I'd check `NIX_SSL_CERT_FILE` only after the canonical file locations have been checked. This also prevents you from abusing this environment variable to inject another bundle.
* `NIX_SSL_CERT_FILE` needs to be respected on all platforms except Windows since you can install and use `nix` on Darwin and FreeBSD as well (I think OpenBSD is not supported atm, but it may be in the future).

**Comment by hannesm** [comment]
Thanks for your quick response. The underlying question is the failure semantics.

> I think two things can be improved about this:
>    I'd check NIX_SSL_CERT_FILE only after the canonical file locations have been checked.

Well, and it would lead to very strange semantics depending on:
-  what is set (and available) in your build sandbox!?
- is the file where NIX_SSL_CERT_FILE points to readable?

>    NIX_SSL_CERT_FILE needs to be respected on all platforms except Windows since you can install and use nix on Darwin and FreeBSD as well (I think OpenBSD is not supported atm, but it may be in the future).

Ah, ok. Good point!

The behaviour with this (or a revised patch) is IMHO deterministic:
- if NIX_SSL_CERT_FILE is set, use the file which this points to (and fail if the file does not exist + print a message which path is used)
- if NIX_SSL_CERT_FILE is not set, use the existing detection mechanisms

What you propose would not guarantee that NIX_SSL_CERT_FILE would be used if set (but only if no other trust anchor would be available). Which results in unreproducible behaviour depending on the environment and file system. Since I have not used Nix, I'm not sure about their desired semantics.

> This also prevents you from abusing this environment variable to inject another bundle.

Well, fair enough -- but isn't that what you actually want "a reliable way to inject another bundle"?

Long story short, I'm still very unconvinced of the NixOS approach to use global mutable state (environment variables) for "sandbox / reproducible builds".

**Comment by sternenseemann** [comment]
Your approach is fine as well, my suggestion was also reflecting your initial reluctance to support an environment variable.

Checking the environment variable after the canonical location would introduce the problem of unreproducible behavior in some cases since there are two kinds of nix builds essentially:

* sandboxed builds: Here only a minimal, fake file system is available and most notably `/etc/ssl` wouldn't exist for the build process. In this case it would be absolutely fine to check the environment variable after the canonical locations
* unsandboxed builds unfortunately also exist: a) for impure builds using system toolchains (mostly on macOS for XCode related stuff) and b) if the user doesn't have enough permissions (is a trusted user) sandboxed builds are also disabled (this has something to do with the way the nix sandbox works and being a trusted user gives you effective root iirc).

So in conclusion, checking `NIX_SSL_CERT_FILE` first avoids some edge cases and probably also keeps the code a bit cleaner (I've had a look at it and it seems to be really annoying to match the failure case and do something alternatively with `Rresult`).

> is the file where NIX_SSL_CERT_FILE points to readable?

It being set and not readable in the build environment would be a nix / nixpkgs bug.

**Review by sternenseemann — APPROVED** [review]

**Comment by sternenseemann** [comment]
Thanks!
